File: scripts/normalize_posts.py
#!/usr/bin/env python3
"""
Normalize posts in _posts/: ensure `lang`, `ref`, and `permalink` are present.
Usage: python scripts/normalize_posts.py

Behavior:
- Parses each Markdown file in `_posts/` with YAML front matter.
- If missing `lang`, attempts a heuristic language detection (accents + stopwords).
- If missing `ref`, generates one from `title` or filename (slugified).
- If missing `permalink`, generates a permalink:
  - English posts: `/en/YYYY/MM/DD/slug/`
  - Portuguese posts: `/YYYY/MM/DD/slug/`
- Writes updated front matter back to file, preserving existing content.

This script is conservative: it only adds missing keys and avoids reformatting complex YAML values.
"""

import os
import re
import sys
import unicodedata
from datetime import datetime

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
POSTS_DIR = os.path.join(ROOT, '_posts')

PT_WORDS = [' o ', ' a ', ' de ', ' que ', ' não ', 'é ', ' com ', ' por ', ' para ', 'um ', 'uma ']
EN_WORDS = [' the ', ' and ', ' of ', ' to ', ' is ', ' in ', ' that ', ' a ', ' an ']
ACCENTS = 'áàãâéèêíìîóòõôúùûç'

FRONT_MATTER_RE = re.compile(r'^---\s*\n(.*?)\n---\s*\n', re.S)
KEY_VALUE_RE = re.compile(r'^([A-Za-z0-9_\-]+):\s*(.*)$')


def slugify(text):
    text = str(text)
    text = unicodedata.normalize('NFKD', text)
    text = text.encode('ascii', 'ignore').decode('ascii')
    text = text.lower()
    text = re.sub(r'[^a-z0-9\s-]', '', text)
    text = re.sub(r'\s+', '-', text).strip('-')
    return text or 'post'


def detect_lang(text):
    t = ' ' + text.lower() + ' '
    # quick accent check
    if any(ch in t for ch in ACCENTS):
        return 'pt'
    pt_score = sum(t.count(w) for w in PT_WORDS)
    en_score = sum(t.count(w) for w in EN_WORDS)
    return 'pt' if pt_score >= en_score else 'en'


def parse_front_matter(text):
    m = FRONT_MATTER_RE.match(text)
    if not m:
        return None, text
    fm = m.group(1)
    rest = text[m.end():]
    lines = fm.splitlines()
    data = {}
    raw_lines = []
    for line in lines:
        raw_lines.append(line)
        kv = KEY_VALUE_RE.match(line)
        if kv:
            k = kv.group(1).strip()
            v = kv.group(2).strip()
            # remove surrounding quotes
            if (v.startswith('"') and v.endswith('"')) or (v.startswith("'") and v.endswith("'")):
                v = v[1:-1]
            data[k] = v
    return data, rest


def build_front_matter(data, original_raw_lines):
    # preserve original order where possible; insert new keys after existing ones
    keys = [line.split(':',1)[0] for line in original_raw_lines if ':' in line]
    used = set()
    out_lines = []
    for k in keys:
        k = k.strip()
        if k in data:
            out_lines.append(f"{k}: {data[k]}")
            used.add(k)
        else:
            # keep original raw line if unknown
            for line in original_raw_lines:
                if line.startswith(k + ':'):
                    out_lines.append(line)
                    break
    # append any remaining keys
    for k, v in data.items():
        if k not in used:
            out_lines.append(f"{k}: {v}")
    return '---\n' + '\n'.join(out_lines) + '\n---\n\n'


def process_post(path):
    with open(path, 'r', encoding='utf-8') as f:
        text = f.read()
    fm, body = parse_front_matter(text)
    if fm is None:
        print(f"Skipping (no front matter): {path}")
        return False

    original_fm_block = FRONT_MATTER_RE.match(text).group(1).splitlines()

    changed = False

    title = fm.get('title') or ''
    filename = os.path.basename(path)

    # ensure lang
    if 'lang' not in fm or not fm['lang'].strip():
        # detect from title + body
        sample = (title + '\n' + body[:2000])
        detected = detect_lang(sample)
        fm['lang'] = detected
        print(f"Added lang: {detected} -> {path}")
        changed = True

    # ensure ref
    if 'ref' not in fm or not fm['ref'].strip():
        # prefer explicit slug from title or filename
        base = title or re.sub(r'^\d{4}-\d{2}-\d{2}-', '', filename)
        r = slugify(base)
        fm['ref'] = r
        print(f"Added ref: {r} -> {path}")
        changed = True

    # ensure permalink
    if 'permalink' not in fm or not fm['permalink'].strip():
        # determine date parts
        date_str = fm.get('date') or filename[0:10]
        year = month = day = None
        try:
            dt = datetime.fromisoformat(date_str.strip())
            year = dt.year
            month = f"{dt.month:02d}"
            day = f"{dt.day:02d}"
        except Exception:
            # try parse common Jekyll format
            m = re.match(r'^(\d{4})-(\d{2})-(\d{2})', filename)
            if m:
                year, month, day = m.group(1), m.group(2), m.group(3)
        slug = fm.get('slug') or slugify(title or re.sub(r'^\d{4}-\d{2}-\d{2}-', '', filename))
        if fm.get('lang') == 'en':
            if year and month and day:
                permalink = f"/en/{year}/{month}/{day}/{slug}/"
            else:
                permalink = f"/en/{slug}/"
        else:
            if year and month and day:
                permalink = f"/{year}/{month}/{day}/{slug}/"
            else:
                permalink = f"/{slug}/"
        fm['permalink'] = permalink
        print(f"Added permalink: {permalink} -> {path}")
        changed = True

    if changed:
        new_fm_block = build_front_matter(fm, original_fm_block)
        new_text = new_fm_block + body
        with open(path, 'w', encoding='utf-8') as f:
            f.write(new_text)
    return changed


def main():
    if not os.path.isdir(POSTS_DIR):
        print(f"_posts directory not found at {POSTS_DIR}")
        return
    files = [f for f in os.listdir(POSTS_DIR) if f.endswith('.md') or f.endswith('.markdown')]
    if not files:
        print("No posts found")
        return
    changed_any = False
    for fn in files:
        path = os.path.join(POSTS_DIR, fn)
        try:
            changed = process_post(path)
            if changed:
                changed_any = True
        except Exception as e:
            print(f"Error processing {path}: {e}")
    if changed_any:
        print("Normalization complete: some files were modified.")
    else:
        print("All posts already normalized.")


if __name__ == '__main__':
    main()

File: _site/scripts/normalize_posts.js
#!/usr/bin/env node
import fs from 'fs/promises';
import path from 'path';

const ROOT = path.resolve(new URL('.', import.meta.url).pathname, '..');
const POSTS_DIR = path.join(process.cwd(), '_posts');

const PT_WORDS = [' o ', ' a ', ' de ', ' que ', ' não ', 'é ', ' com ', ' por ', ' para ', 'um ', 'uma '];
const EN_WORDS = [' the ', ' and ', ' of ', ' to ', ' is ', ' in ', ' that ', ' a ', ' an '];
const ACCENTS_RE = /[áàãâéèêíìîóòõôúùûçÁÀÃÂÉÈÊÍÌÎÓÒÕÔÚÙÛÇ]/;

const FRONT_MATTER_RE = /^---\s*\n([\s\S]*?)\n---\s*\n/;
const KEY_VALUE_RE = /^([A-Za-z0-9_\-]+):\s*(.*)$/;

function slugify(text){
  if(!text) return 'post';
  const normalized = text.normalize('NFKD').replace(/\p{Diacritic}/gu, '');
  return normalized
    .toString()
    .toLowerCase()
    .replace(/[^a-z0-9\s-]/g, '')
    .trim()
    .replace(/\s+/g, '-');
}

function detectLang(sample){
  const t = (' ' + (sample || '').toLowerCase() + ' ');
  if(ACCENTS_RE.test(sample)) return 'pt';
  const ptScore = PT_WORDS.reduce((s,w)=>s + (t.split(w).length -1), 0);
  const enScore = EN_WORDS.reduce((s,w)=>s + (t.split(w).length -1), 0);
  return ptScore >= enScore ? 'pt' : 'en';
}

function parseFrontMatter(text){
  const m = FRONT_MATTER_RE.exec(text);
  if(!m) return {fm: null, rawLines: [], body: text};
  const fmBlock = m[1];
  const rest = text.slice(m[0].length);
  const lines = fmBlock.split(/\r?\n/);
  const data = {};
  for(const line of lines){
    const kv = KEY_VALUE_RE.exec(line);
    if(kv){
      let k = kv[1].trim();
      let v = kv[2].trim();
      if((v.startsWith('"') && v.endsWith('"')) || (v.startsWith("'") && v.endsWith("'"))){
        v = v.slice(1,-1);
      }
      data[k] = v;
    }
  }
  return {fm: data, rawLines: lines, body: rest};
}

function buildFrontMatter(data, originalRawLines){
  const keysInOrder = [];
  for(const line of originalRawLines){
    const idx = line.indexOf(':');
    if(idx !== -1){
      keysInOrder.push(line.slice(0, idx).trim());
    }
  }
  const used = new Set();
  const out = [];
  for(const k of keysInOrder){
    if(k in data){
      out.push(`${k}: ${data[k]}`);
      used.add(k);
    } else {
      // keep original raw line
      for(const line of originalRawLines){
        if(line.startsWith(k + ':')){ out.push(line); break; }
      }
    }
  }
  for(const k of Object.keys(data)){
    if(!used.has(k)) out.push(`${k}: ${data[k]}`);
  }
  return '---\n' + out.join('\n') + '\n---\n\n';
}

async function processPost(filePath){
  const filename = path.basename(filePath);
  let text = await fs.readFile(filePath, 'utf8');
  const {fm, rawLines, body} = parseFrontMatter(text);
  if(!fm){
    console.log(`Skipping (no front matter): ${filePath}`);
    return false;
  }
  let changed = false;

  const title = fm.title || '';

  if(!fm.lang || !fm.lang.trim()){
    const sample = (title + '\n' + body.slice(0,2000));
    const detected = detectLang(sample);
    fm.lang = detected;
    console.log(`Added lang: ${detected} -> ${filePath}`);
    changed = true;
  }

  if(!fm.ref || !fm.ref.trim()){
    const base = title || filename.replace(/^\d{4}-\d{2}-\d{2}-/, '');
    const r = slugify(base);
    fm.ref = r;
    console.log(`Added ref: ${r} -> ${filePath}`);
    changed = true;
  }

  if(!fm.permalink || !fm.permalink.trim()){
    let year, month, day;
    const dateStr = fm.date || filename.slice(0,10);
    const d = new Date(dateStr);
    if(!isNaN(d.getTime())){
      year = d.getFullYear();
      month = String(d.getMonth()+1).padStart(2,'0');
      day = String(d.getDate()).padStart(2,'0');
    } else {
      const m = filename.match(/^(\d{4})-(\d{2})-(\d{2})/);
      if(m){ year = m[1]; month = m[2]; day = m[3]; }
    }
    const slug = fm.slug || slugify(title || filename.replace(/^\d{4}-\d{2}-\d{2}-/, ''));
    let permalink;
    if(fm.lang === 'en'){
      if(year && month && day) permalink = `/en/${year}/${month}/${day}/${slug}/`;
      else permalink = `/en/${slug}/`;
    } else {
      if(year && month && day) permalink = `/${year}/${month}/${day}/${slug}/`;
      else permalink = `/${slug}/`;
    }
    fm.permalink = permalink;
    console.log(`Added permalink: ${permalink} -> ${filePath}`);
    changed = true;
  }

  if(changed){
    const newFm = buildFrontMatter(fm, rawLines);
    const newText = newFm + body;
    await fs.writeFile(filePath, newText, 'utf8');
  }
  return changed;
}

async function main(){
  try{
    const stat = await fs.stat(POSTS_DIR);
    if(!stat.isDirectory()){
      console.error(`_posts is not a directory: ${POSTS_DIR}`);
      return;
    }
  }catch(e){
    console.error(`_posts directory not found at ${POSTS_DIR}`);
    return;
  }
  const entries = await fs.readdir(POSTS_DIR);
  const files = entries.filter(f => f.endsWith('.md') || f.endsWith('.markdown'));
  if(files.length === 0){ console.log('No posts found'); return; }
  let changedAny = false;
  for(const fn of files){
    const p = path.join(POSTS_DIR, fn);
    try{
      const changed = await processPost(p);
      if(changed) changedAny = true;
    }catch(err){
      console.error(`Error processing ${p}: ${err}`);
    }
  }
  if(changedAny) console.log('Normalization complete: some files were modified.');
  else console.log('All posts already normalized.');
}

if(require.main === module){
  main();
}
